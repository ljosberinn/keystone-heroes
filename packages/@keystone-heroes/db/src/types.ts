export {
  Affixes,
  Covenants,
  EventType,
  PlayableClass,
  Prisma,
  Role,
  SpecName,
} from "@prisma/client";
export type {
  Ability,
  Affix,
  Character,
  Class,
  Conduit,
  Covenant,
  CovenantTrait,
  Dungeon,
  Event,
  Expansion,
  Fight,
  Legendary,
  NPC,
  Player,
  PlayerConduit,
  PlayerCovenantTrait,
  PlayerFight,
  PlayerTalent,
  Pull,
  PullNPC,
  PullZone,
  Region,
  Report,
  Season,
  Server,
  Soulbind,
  Spec,
  Talent,
  WCLAuth,
  Week,
  Zone,
} from "@prisma/client";
