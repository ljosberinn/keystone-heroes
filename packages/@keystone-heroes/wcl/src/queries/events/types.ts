export type BaseEvent<T extends Record<string, unknown>> = T & {
  timestamp: number;
  // pin: "0" but irrelevant
};

type EncounterStartEvent = BaseEvent<{
  type: "encounterstart";
  encounterID: number;
  name: string;
  difficulty: number;
  size: number;
  level: number;
  affixes: number[];
}>;

type CombatantInfoEvent = BaseEvent<{
  type: "combatantinfo";
  sourceID: number;
  gear: [];
  auras: {
    source: number;
    ability: number;
    stacks: number;
    icon: string;
    name?: string;
  }[];
  expansion: string;
  faction: number;
  specID: number;
  covenantID: number;
  soulbindID: number;
  strength: number;
  agility: number;
  stamina: number;
  intellect: number;
  dodge: number;
  parry: number;
  block: number;
  armor: number;
  critMelee: number;
  critRanged: number;
  critSpell: number;
  speed: number;
  leech: number;
  hasteMelee: number;
  hasteRanged: number;
  hasteSpell: number;
  avoidance: number;
  mastery: number;
  versatilityDamageDone: number;
  versatilityHealingDone: number;
  versatilityDamageReduction: number;
  talents: { id: number; icon: string }[];
  pvpTalents: { id: number; icon: string }[];
  artifact: { traitID: number; rank: number; spellID: number; icon: string }[];
  heartOfAzeroth: {
    traitID: number;
    rank: number;
    spellID: number;
    icon: string;
  }[];
}>;

export type ApplyDebuffEvent = BaseEvent<{
  type: "applydebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  targetInstance?: number;
  targetMarker?: number;
}>;

export type ApplyDebuffStackEvent = BaseEvent<{
  type: "applydebuffstack";
  stack: number;
  abilityGameID: number;
}>;

export type ApplyBuffEvent = BaseEvent<{
  type: "applybuff";
  sourceID: number;
  sourceInstance?: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
}>;

export type RemoveBuffEvent = BaseEvent<{
  type: "removebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  sourceMarker: number;
  targetMarker: number;
}>;

export type CastEvent = BaseEvent<{
  type: "cast";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  sourceMarker?: number;
}>;

export type DamageEvent = BaseEvent<{
  type: "damage";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  hitType: number;
  amount: number;
  mitigated?: number;
  unmitigatedAmount?: number;
  tick?: boolean;
  sourceMarker?: number;
  absorbed?: number;
  overkill?: number;
  buffs?: string;
}>;

export type BeginCastEvent = BaseEvent<{
  type: "begincast";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
}>;

export type RemoveDebuffEvent = BaseEvent<{
  type: "removedebuff";
  sourceID: number;
  targetID: number;
  targetInstance: number;
  abilityGameID: number;
}>;

type SummonEvent = BaseEvent<{
  type: "summon";
  sourceID: number;
  targetID: number;
  targetInstance: number;
  abilityGameID: number;
}>;

type PhaseStartEvent = BaseEvent<{
  type: "phasestart";
  encounterID: number;
  name: string;
  difficulty: number;
  size: number;
}>;

export type HealEvent = BaseEvent<{
  type: "heal";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  hitType: number;
  amount: number;
  sourceMarker?: number;
  targetMarker?: number;
  tick?: boolean;
  overheal?: number;
  absorbed?: number;
}>;

type EnergizeEvent = BaseEvent<{
  type: "energize";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  resourceChange: number;
  resourceChangeType: number;
  otherResourceChange: number;
  waste: number;
  sourceMarker?: number;
  targetMarker?: number;
}>;

export type ApplyBuffStackEvent = BaseEvent<{
  type: "applybuffstack";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  stack: number;
  sourceMarker?: number;
  targetMarker?: number;
}>;

export type InterruptEvent = BaseEvent<{
  type: "interrupt";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  extraAbilityGameID: number;
  sourceMarker?: number;
}>;

export type AbsorbEvent = BaseEvent<{
  type: "absorbed";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  attackerID: number;
  amount: number;
  extraAbilityGameID: number;
  sourceMarker: number;
  targetMarker: number;
}>;

export type RefreshDebuffEvent = BaseEvent<{
  type: "refreshdebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
}>;

type DispelEvent = BaseEvent<{
  type: "dispel";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  extraAbilityGameID: number;
  isBuff: boolean;
  sourceMarker: number;
  targetMarker: number;
}>;

export type DeathEvent = BaseEvent<{
  type: "death";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  // only present if NPC kills player
  killerID?: number;
  killerInstance?: number;
  killingAbilityGameID?: number;
  targetMarker?: number;
}>;

export type AnyEvent =
  | DeathEvent
  | DispelEvent
  | RefreshDebuffEvent
  | AbsorbEvent
  | InterruptEvent
  | ApplyBuffStackEvent
  | EnergizeEvent
  | HealEvent
  | PhaseStartEvent
  | SummonEvent
  | RemoveDebuffEvent
  | BeginCastEvent
  | DamageEvent
  | CastEvent
  | RemoveBuffEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
  | ApplyDebuffEvent
  | CombatantInfoEvent
  | EncounterStartEvent;

export type AllTrackedEventTypes = (
  | CastEvent
  | DeathEvent
  | AbsorbEvent
  | DamageEvent
  | HealEvent
  | InterruptEvent
  | BeginCastEvent
  | ApplyBuffEvent
  | ApplyDebuffEvent
  | ApplyDebuffStackEvent
  | RemoveBuffEvent
  | ApplyBuffStackEvent
)[];
