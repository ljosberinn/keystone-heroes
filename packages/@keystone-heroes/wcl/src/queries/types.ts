import type { PlayableClass, Role, SpecName } from "@prisma/client";

export type Table = {
  totalTime: number;
  itemLevel: number;
  logVersion: number;
  gameVersion: number;
  composition: Composition[];
  damageDone: DamageDone[];
  healingDone: HealingDone[];
  damageTaken: DamageTaken[];
  deathEvents: DeathEvent[];
  playerDetails: PlayerDetails;
};

type Composition = {
  name: string;
  id: number;
  guid: number;
  type: PlayableClass;
  specs: SpecMeta[];
};

type SpecMeta = {
  spec: SpecName;
  role: Role;
};

export type DamageDone = Pick<Composition, "name" | "id" | "guid"> & {
  total: number;
  icon: string;
  type: Composition["type"] | "Pet" | "NPC";
};

export type HealingDone = DamageDone;

export type DamageTaken = Pick<DamageDone, "name" | "guid" | "total"> & {
  abilityIcon: string;
  type: number;
};

type DeathEvent = Pick<DamageDone, "name" | "id" | "guid" | "type" | "icon"> & {
  deathTime: number;
  ability: Omit<DamageTaken, "total">;
};

export type PlayerDetails = {
  tanks: InDepthCharacterInformation[];
  dps: InDepthCharacterInformation[];
  healers: InDepthCharacterInformation[];
};

export type InDepthCharacterInformation = {
  name: string;
  id: number;
  guid: number;
  type: PlayableClass;
  server: string;
  icon: string;
  specs: SpecName[];
  minItemLevel: number;
  maxItemLevel: number;
  combatantInfo: CombatantInfo;
};

type CombatantInfo = {
  stats: Stats;
  talents: Talent[];
  gear: (Item | LegendaryItem)[];
  /**
   * @deprecated https://canary.discord.com/channels/180033360939319296/681904912090529801/863865520486088704
   */
  artifact?: CustomPowerSet[];
  /**
   * @deprecated https://canary.discord.com/channels/180033360939319296/681904912090529801/863865520486088704
   */
  heartOfAzeroth?: SecondaryCustomPowerSet[];
  customPowerSet: CustomPowerSet[];
  secondaryCustomPowerSet: SecondaryCustomPowerSet[];
  tertiaryCustomPowerSet: unknown[];
  specIDs: number[];
  factionID: number;
  covenantID?: number;
  soulbindID?: number;
};

type Stats = {
  Speed: Stat;
  Dodge: Stat;
  Mastery: Stat;
  Stamina: Stat;
  Haste: Stat;
  Leech: Stat;
  Armor: Stat;
  Agility: Stat;
  Crit: Stat;
  "Item Level": Stat;
  Parry: Stat;
  Avoidance: Stat;
  Versatility: Stat;
  Intellect: Stat;
};

type Stat = {
  min: number;
  max: number;
};

export type Talent = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
};

type CustomPowerSet = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
  total: number;
};

export enum ItemQuality {
  POOR = 1,
  COMMON = 2,
  RARE = 3,
  SUPERIOR = 4,
  LEGENDARY = 5,
}

export type Item = {
  id: number;
  slot: number;
  quality: ItemQuality;
  icon: string;
  name?: string;
  itemLevel: number;
  bonusIDs?: number[];
  gems?: Gem[];
  permanentEnchant?: number;
  permanentEnchantName?: string;
  onUseEnchant?: number;
  onUseEnchantName?: string;
  effectID?: number;
  effectIcon?: string;
  effectName?: string;
  temporaryEnchant?: number;
  temporaryEnchantName?: string;
};

export type LegendaryItem = Item & {
  quality: ItemQuality.LEGENDARY;
  effectID: number;
  effectIcon: string;
  effectName: string;
};

type Gem = {
  id: number;
  itemLevel: number;
  icon: string;
};

type SecondaryCustomPowerSet = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
  total: number;
};

export type Conduit = SecondaryCustomPowerSet;
export type CovenantTrait = CustomPowerSet;
