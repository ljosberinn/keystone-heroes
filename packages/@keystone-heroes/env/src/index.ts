/* eslint-disable @typescript-eslint/no-non-null-assertion */
/** ********************
 * utilities
 *********************/
export const IS_BROWSER = typeof window !== "undefined";
export const IS_PRODUCTION = process.env.NODE_ENV === "production";
export const IS_DEVELOPMENT = process.env.NODE_ENV === "development";
export const IS_TEST = process.env.NODE_ENV === "test";

/** *********************
 * meta
 *********************/
export const COMMIT_SHA = process.env.NEXT_PUBLIC_COMMIT_SHA;
export const BUILD_TIME = process.env.NEXT_PUBLIC_BUILD_TIME!;
export const BUILD_TIMESTAMP = process.env.NEXT_PUBLIC_BUILD_TIMESTAMP!;

/** *********************
 * WCL
 *********************/
export const WCL_OAUTH_ENDPOINT = "https://www.warcraftlogs.com/oauth/token";
export const WCL_GQL_ENDPOINT = "https://www.warcraftlogs.com/api/v2/client";
export const WCL_CLIENT_ID = process.env.WCL_CLIENT_ID!;
export const WCL_CLIENT_SECRET = process.env.WCL_CLIENT_SECRET!;

/** *********************
 * Battle.net
 *********************/
export const BATTLENET_CLIENT_ID = process.env.BATTLENET_CLIENT_ID!;
export const BATTLENET_CLIENT_SECRET = process.env.BATTLENET_CLIENT_SECRET!;

/** *********************
 * general
 *********************/
export const MIN_KEYSTONE_LEVEL = 10;
