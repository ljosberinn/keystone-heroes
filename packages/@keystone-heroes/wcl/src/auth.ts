import { prisma } from "@keystone-heroes/db/prisma";
import type { WCLAuth } from "@keystone-heroes/db/types";

export type WCLOAuthResponse = {
  access_token: string;
  expires_in: number;
  token_type: "Bearer";
};

export const setWCLAuthentication = async ({
  access_token,
  expires_in,
}: WCLOAuthResponse): Promise<void> => {
  const payload: Omit<WCLAuth, "id"> = {
    token: access_token,
    expiresAt: Math.round(Date.now() / 1000) + expires_in,
  };

  await prisma.wCLAuth.upsert({
    create: payload,
    where: {
      id: 1,
    },
    update: payload,
  });
};

export const getWCLAuthentication = (): Promise<WCLAuth | null> => {
  if (process.env.NODE_ENV === "test") {
    return Promise.resolve({
      id: 1,
      token: "mock-token",
      expiresAt: Date.now() + 28 * 24 * 60 * 60 * 1000,
    });
  }

  return prisma.wCLAuth.findFirst();
};
