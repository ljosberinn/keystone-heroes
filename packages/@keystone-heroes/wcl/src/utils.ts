export const isValidReportId = (id?: string | string[]): id is string =>
  !Array.isArray(id) && id?.length === 16 && !id.includes(".");

export const ONGOING_REPORT_THRESHOLD = 18 * 60 * 60 * 1000;

/**
 * assume a report may still be ongoing if its less than one day old
 */
export const maybeOngoingReport = (endTime: number): boolean =>
  ONGOING_REPORT_THRESHOLD > Date.now() - endTime;
