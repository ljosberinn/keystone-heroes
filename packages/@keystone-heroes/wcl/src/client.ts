import {
  WCL_CLIENT_ID,
  WCL_CLIENT_SECRET,
  WCL_GQL_ENDPOINT,
  WCL_OAUTH_ENDPOINT,
} from "@keystone-heroes/env";
import { GraphQLClient } from "graphql-request";

import type { WCLOAuthResponse } from "./auth";
import { getWCLAuthentication, setWCLAuthentication } from "./auth";
import { getSdk } from "./types";
import type { Sdk } from "./types";

type Cache = {
  sdk: Sdk | null;
  client: GraphQLClient | null;
  expiresAt: number | null;
};

const cache: Cache = {
  sdk: null,
  client: null,
  expiresAt: null,
};

export const getCachedSdk = async (): Promise<Sdk> => {
  if (cache.sdk) {
    return cache.sdk;
  }

  const client = await getGqlClient();
  cache.sdk = cache.sdk ?? getSdk(client);

  return cache.sdk;
};

export const getGqlClient = async (): Promise<GraphQLClient> => {
  if (
    cache.client &&
    cache.expiresAt &&
    cache?.expiresAt > Date.now() + 60 * 1000
  ) {
    return cache.client;
  }

  const cached = await getWCLAuthentication();

  if (cached?.token && cached?.expiresAt && !cache.client && !cache.expiresAt) {
    cache.client = new GraphQLClient(WCL_GQL_ENDPOINT, {
      headers: {
        authorization: `Bearer ${cached.token}`,
      },
    });

    cache.expiresAt = cached.expiresAt * 1000;

    return cache.client;
  }

  try {
    const body = new URLSearchParams({
      client_id: WCL_CLIENT_ID,
      client_secret: WCL_CLIENT_SECRET,
      grant_type: "client_credentials",
    }).toString();

    const response = await fetch(WCL_OAUTH_ENDPOINT, {
      body,
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
    });

    if (response.ok) {
      const json: WCLOAuthResponse = await response.json();

      await setWCLAuthentication(json);

      cache.client =
        cache.client ??
        new GraphQLClient(WCL_GQL_ENDPOINT, {
          headers: {
            authorization: `Bearer ${json.access_token}`,
          },
        });

      cache.expiresAt = cache.expiresAt ?? Date.now() + json.expires_in;

      return cache.client;
    }

    throw new Error("unable to authenticate with WCL");
  } catch {
    throw new Error("unable to authenticate with WCL");
  }
};
