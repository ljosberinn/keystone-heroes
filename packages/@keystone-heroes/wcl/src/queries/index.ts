import { getCachedSdk } from "../client";
import type {
  FightPullsQuery,
  FightPullsQueryVariables,
  InitialReportDataQuery,
  InitialReportDataQueryVariables,
  TableQuery,
  TableQueryVariables,
} from "../types";
import type { Table } from "./table";

export * from "./events";

export type {
  Conduit,
  DamageDone,
  DamageTaken,
  HealingDone,
  InDepthCharacterInformation,
  Item,
  LegendaryItem,
  PlayerDetails,
  CovenantTrait,
  Table,
  Talent,
} from "./table";
export { ItemQuality } from "./table";

export const getInitialReportData = async (
  params: InitialReportDataQueryVariables
): Promise<InitialReportDataQuery> => {
  const sdk = await getCachedSdk();

  return sdk.InitialReportData(params);
};

// TODO: duplicate
type DeepNullablePath<T, P> = P extends []
  ? T
  : P extends [infer A, ...infer B]
  ? DeepNullablePath<NonNullable<T>[A & keyof NonNullable<T>] | undefined, B>
  : never;

type TypedTableQuery = {
  reportData?: {
    report?:
      | (Omit<
          DeepNullablePath<TableQuery, ["reportData", "report"]>,
          "table"
        > & { table?: { data: Table } })
      | null;
  } | null;
};

export const getTableData = async (
  params: TableQueryVariables
): Promise<TypedTableQuery> => {
  const sdk = await getCachedSdk();

  return sdk.Table(params);
};

export const getFightPulls = async (
  params: FightPullsQueryVariables
): Promise<FightPullsQuery> => {
  const sdk = await getCachedSdk();

  return sdk.FightPulls(params);
};
