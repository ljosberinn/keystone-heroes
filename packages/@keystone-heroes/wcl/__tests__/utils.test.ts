import { graphql } from "msw";
import { setupServer } from "msw/node";

import type { DamageEvent, HealEvent } from "../src/queries/events/types";
import {
  createChunkByThresholdReducer,
  createEventFetcher,
  getEvents,
  reduceEventsByPlayer,
} from "../src/queries/events/utils";
import { EventDataType, HostilityType } from "../src/types";
import NPCDamagesPlayerEvents from "./fixtures/NPCDamagesPlayerEvents.json";
import kyrianOrbHealEvents from "./fixtures/kyrianOrbHealEvents.json";
import playerDamagesNPCEvents from "./fixtures/playerDamagesNPCEvents.json";

describe("getEvents", () => {
  const server = setupServer();

  beforeAll(() => {
    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  test("recursively requests events until nextPageTimestamp is absent", async () => {
    let requestNumber = 0;

    server.use(
      graphql.query("EventData", (req, res, ctx) => {
        expect(req.variables.startTime).toBe(
          // eslint-disable-next-line jest/no-if
          requestNumber === 0 ? 0 : requestNumber * 1000
        );

        requestNumber++;

        // eslint-disable-next-line jest/no-if
        if (requestNumber === 3) {
          return res(
            ctx.data({
              reportData: {
                report: {
                  events: {
                    data: [requestNumber],
                  },
                },
              },
            })
          );
        }

        return res(
          ctx.data({
            reportData: {
              report: {
                events: {
                  data: [requestNumber],
                  nextPageTimestamp: requestNumber * 1000,
                },
              },
            },
          })
        );
      })
    );

    const response = await getEvents({
      dataType: EventDataType.DamageDone,
      hostilityType: HostilityType.Enemies,
      reportID: "",
      startTime: 0,
      endTime: 1,
    });

    expect(response).toStrictEqual([1, 2, 3]);
    expect(requestNumber).toBe(3);
  });

  test("requests only once if nextPageTimestamp is absent initially", async () => {
    const requestNumber = 1;
    const dummy = jest.fn();

    server.use(
      graphql.query("EventData", (_req, res, ctx) => {
        if (requestNumber !== 1) {
          dummy();

          return res(
            ctx.data({
              reportData: {
                report: {
                  events: {
                    data: [requestNumber],
                  },
                },
              },
            })
          );
        }

        return res(
          ctx.data({
            reportData: {
              report: {
                events: {
                  data: [requestNumber],
                },
              },
            },
          })
        );
      })
    );

    const response = await getEvents({
      dataType: EventDataType.DamageDone,
      hostilityType: HostilityType.Enemies,
      reportID: "",
      startTime: 0,
      endTime: 1,
    });

    expect(response).toStrictEqual([1]);
    expect(requestNumber).toBe(1);
    expect(dummy).not.toHaveBeenCalled();
  });

  test("fails gracefully given no response", async () => {
    server.use(
      graphql.query("EventData", (_req, res, ctx) => {
        return res(ctx.data({}));
      })
    );

    const response = await getEvents({
      dataType: EventDataType.DamageDone,
      hostilityType: HostilityType.Enemies,
      reportID: "",
      startTime: 0,
      endTime: 1,
    });

    expect(response).toStrictEqual([]);
  });

  test("fails gracefully given an invalid response", async () => {
    server.use(
      graphql.query("EventData", (_req, res, ctx) => {
        return res(ctx.data({ foo: "bar" }));
      })
    );

    const response = await getEvents({
      dataType: EventDataType.DamageDone,
      hostilityType: HostilityType.Enemies,
      reportID: "",
      startTime: 0,
      endTime: 1,
    });

    expect(response).toStrictEqual([]);
  });
});

describe("createEventFetcher", () => {
  const server = setupServer();

  beforeAll(() => {
    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  test("forwards params to returned function", async () => {
    const initialParams: Parameters<typeof createEventFetcher>[0] = {
      dataType: EventDataType.DamageDone,
      hostilityType: HostilityType.Enemies,
    };

    const distinguishedParams: Parameters<
      ReturnType<typeof createEventFetcher>
    >[0] = {
      reportID: "",
      startTime: 0,
      endTime: 1,
    };

    server.use(
      graphql.query("EventData", (req, res, ctx) => {
        expect(req.variables).toMatchObject({
          ...initialParams,
          ...distinguishedParams,
        });

        return res(ctx.data({}));
      })
    );

    await createEventFetcher(initialParams)(distinguishedParams);
  });
});

describe("reduceEventsByPlayer", () => {
  describe("DamageEvents", () => {
    test("targetID === damageTaken by player", () => {
      const events: DamageEvent[] = NPCDamagesPlayerEvents.map((event) => ({
        ...event,
        type: "damage",
      }));

      expect(reduceEventsByPlayer(events, "targetID")).toMatchSnapshot();
    });

    test("sourceID === damageDone by player", () => {
      const events: DamageEvent[] = playerDamagesNPCEvents.map((event) => ({
        ...event,
        type: "damage",
      }));

      expect(reduceEventsByPlayer(events, "sourceID")).toMatchSnapshot();
    });
  });

  describe("HealEvents", () => {
    test("targetID === healing received by target", () => {
      const events: HealEvent[] = kyrianOrbHealEvents.map((event) => ({
        ...event,
        type: "heal",
      }));

      expect(reduceEventsByPlayer(events, "targetID")).toMatchSnapshot();
    });

    test("sourceID === healingDone by source", () => {
      const events: HealEvent[] = kyrianOrbHealEvents.map((event) => ({
        ...event,
        type: "heal",
      }));

      expect(reduceEventsByPlayer(events, "sourceID")).toMatchSnapshot();
    });
  });
});

describe("createChunkByThresholdReducer", () => {
  test("groups events into chunks by given threshold", () => {
    const reducer = createChunkByThresholdReducer(5 * 1000);

    const timestampMap = {
      firstChunk: 0,
      secondChunk: 6000,
      thirdChunk: 12_000,
      fourthChunk: 18_000,
    };

    const events: DamageEvent[] = Array.from({ length: 20 }, (_, index) => {
      const chunk =
        index < 5
          ? timestampMap.firstChunk
          : index < 10
          ? timestampMap.secondChunk
          : index < 15
          ? timestampMap.thirdChunk
          : timestampMap.fourthChunk;

      return {
        timestamp: chunk,
        targetID: 1,
        amount: 1,
        hitType: 1,
        sourceID: 100,
        type: "damage",
        abilityGameID: 0,
      };
    });

    const result = events.reduce<DamageEvent[][]>(reducer, []);

    expect(result).toHaveLength(4);
    expect(result.every((chunk) => chunk.length === 5)).toBe(true);

    expect(
      result[0].every((event) => event.timestamp === timestampMap.firstChunk)
    ).toBe(true);

    expect(
      result[1].every((event) => event.timestamp === timestampMap.secondChunk)
    ).toBe(true);

    expect(
      result[2].every((event) => event.timestamp === timestampMap.thirdChunk)
    ).toBe(true);

    expect(
      result[3].every((event) => event.timestamp === timestampMap.fourthChunk)
    ).toBe(true);
  });
});
