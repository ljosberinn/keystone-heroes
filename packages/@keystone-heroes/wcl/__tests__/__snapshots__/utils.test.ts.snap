// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduceEventsByPlayer DamageEvents sourceID === damageDone by player 1`] = `
Array [
  Object {
    "abilityGameID": 346665,
    "amount": 156575,
    "buffs": "337547.",
    "hitType": 1,
    "mitigated": 106,
    "overheal": undefined,
    "sourceID": 4,
    "sourceMarker": 7,
    "targetID": 13,
    "targetInstance": 1,
    "timestamp": 1773978,
    "type": "damage",
    "unmitigatedAmount": 354,
  },
  Object {
    "abilityGameID": 164812,
    "amount": 45889,
    "buffs": "48518.48517.307185.194223.24858.334131.",
    "hitType": 1,
    "overheal": undefined,
    "sourceID": 8,
    "targetID": 10,
    "targetInstance": 4,
    "timestamp": 1778150,
    "type": "damage",
    "unmitigatedAmount": 575,
  },
  Object {
    "abilityGameID": 188389,
    "amount": 111102,
    "buffs": "307185.334131.",
    "hitType": 1,
    "overheal": undefined,
    "sourceID": 2,
    "targetID": 10,
    "targetInstance": 3,
    "timestamp": 1779085,
    "type": "damage",
    "unmitigatedAmount": 419,
  },
  Object {
    "abilityGameID": 1,
    "amount": 1860,
    "buffs": "307185.24858.334131.",
    "hitType": 1,
    "mitigated": 132,
    "overheal": undefined,
    "sourceID": 24,
    "sourceInstance": 1,
    "targetID": 16,
    "timestamp": 1781633,
    "type": "damage",
    "unmitigatedAmount": 442,
  },
  Object {
    "abilityGameID": 157348,
    "amount": 37570,
    "buffs": "325381.345228.80353.343594.345230.324748.307185.336217.334131.",
    "hitType": 1,
    "overheal": undefined,
    "sourceID": 17,
    "sourceInstance": 1,
    "targetID": 10,
    "targetInstance": 3,
    "timestamp": 1782896,
    "type": "damage",
    "unmitigatedAmount": 1374,
  },
  Object {
    "abilityGameID": 108853,
    "amount": 193176,
    "buffs": "190319.341449.345211.116014.80353.343594.26297.307185.327371.",
    "hitType": 2,
    "overheal": undefined,
    "sourceID": 9,
    "targetID": 15,
    "targetInstance": 1,
    "timestamp": 1784461,
    "type": "damage",
    "unmitigatedAmount": 2410,
  },
]
`;

exports[`reduceEventsByPlayer DamageEvents targetID === damageTaken by player 1`] = `
Array [
  Object {
    "abilityGameID": 1,
    "amount": 461439,
    "hitType": 1,
    "mitigated": 11244,
    "overheal": undefined,
    "sourceID": 10,
    "sourceInstance": 1,
    "targetID": 4,
    "targetMarker": 7,
    "timestamp": 1774843,
    "type": "damage",
    "unmitigatedAmount": 15636,
  },
  Object {
    "abilityGameID": 345211,
    "amount": 20789,
    "buffs": "345211.307185.341449.",
    "hitType": 1,
    "mitigated": 157,
    "overheal": undefined,
    "sourceID": 9,
    "targetID": 9,
    "tick": true,
    "timestamp": 1780293,
    "type": "damage",
    "unmitigatedAmount": 587,
  },
  Object {
    "abilityGameID": 1,
    "amount": 144274,
    "hitType": 1,
    "mitigated": 13199,
    "overheal": undefined,
    "sourceID": 10,
    "sourceInstance": 2,
    "targetID": 24,
    "targetInstance": 2,
    "timestamp": 1781866,
    "type": "damage",
    "unmitigatedAmount": 16164,
  },
  Object {
    "abilityGameID": 321821,
    "amount": 71694,
    "hitType": 1,
    "mitigated": 156,
    "overheal": undefined,
    "sourceID": 13,
    "sourceInstance": 7,
    "targetID": 2,
    "timestamp": 1789874,
    "type": "damage",
    "unmitigatedAmount": 1758,
  },
  Object {
    "abilityGameID": 321821,
    "amount": 33020,
    "hitType": 1,
    "mitigated": 251,
    "overheal": undefined,
    "sourceID": 13,
    "sourceInstance": 7,
    "targetID": 1,
    "targetMarker": 2,
    "timestamp": 1789874,
    "type": "damage",
    "unmitigatedAmount": 1758,
  },
  Object {
    "abilityGameID": 321821,
    "amount": 20916,
    "hitType": 1,
    "mitigated": 170,
    "overheal": undefined,
    "sourceID": 13,
    "sourceInstance": 7,
    "targetID": 8,
    "timestamp": 1789874,
    "type": "damage",
    "unmitigatedAmount": 1758,
  },
  Object {
    "abilityGameID": 1,
    "amount": 889,
    "hitType": 1,
    "mitigated": 2544,
    "overkill": 5051,
    "sourceID": 13,
    "sourceInstance": 8,
    "targetID": 14,
    "targetInstance": 2,
    "timestamp": 1877650,
    "type": "damage",
    "unmitigatedAmount": 8484,
  },
]
`;

exports[`reduceEventsByPlayer HealEvents sourceID === healingDone by source 1`] = `
Array [
  Object {
    "abilityGameID": 344422,
    "amount": 211052,
    "buffs": "216251.333089.307185.52127.335161.",
    "hitType": 1,
    "overheal": 372945,
    "sourceID": 1,
    "sourceMarker": 2,
    "targetID": 1,
    "targetMarker": 2,
    "timestamp": 1913800,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 228526,
    "buffs": "341449.343594.307185.335161.324748.",
    "hitType": 2,
    "overheal": 458486,
    "sourceID": 9,
    "targetID": 9,
    "timestamp": 1915001,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 241688,
    "buffs": "203981.337547.974.61295.335161.324226.",
    "hitType": 1,
    "overheal": 326568,
    "sourceID": 4,
    "sourceMarker": 7,
    "targetID": 4,
    "targetMarker": 7,
    "timestamp": 1916507,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 93184,
    "buffs": "334131.307185.335161.",
    "hitType": 1,
    "overheal": 435872,
    "sourceID": 8,
    "targetID": 8,
    "timestamp": 1923089,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 148533,
    "buffs": "307185.345230.324748.334131.335161.",
    "hitType": 1,
    "overheal": 417429,
    "sourceID": 2,
    "targetID": 2,
    "timestamp": 1924449,
    "type": "heal",
  },
]
`;

exports[`reduceEventsByPlayer HealEvents targetID === healing received by target 1`] = `
Array [
  Object {
    "abilityGameID": 344422,
    "amount": 128882,
    "buffs": "216251.333089.307185.52127.335161.",
    "hitType": 1,
    "overheal": 474213,
    "sourceID": 1,
    "sourceMarker": 2,
    "targetID": 1,
    "targetMarker": 2,
    "timestamp": 1913800,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 152110,
    "buffs": "216251.333089.307185.52127.335161.",
    "hitType": 1,
    "overheal": 456290,
    "sourceID": 1,
    "sourceMarker": 2,
    "targetID": 9,
    "timestamp": 1913816,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 361636,
    "buffs": "216251.333089.307185.52127.335161.",
    "hitType": 1,
    "overheal": 243422,
    "sourceID": 1,
    "sourceMarker": 2,
    "targetID": 4,
    "targetMarker": 7,
    "timestamp": 1913816,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 142701,
    "buffs": "341449.343594.307185.335161.324748.",
    "hitType": 1,
    "overheal": 436731,
    "sourceID": 9,
    "targetID": 2,
    "timestamp": 1915015,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 137654,
    "buffs": "216251.333089.307185.52127.335161.53390.",
    "hitType": 1,
    "overheal": 400644,
    "sourceID": 1,
    "sourceMarker": 2,
    "targetID": 8,
    "timestamp": 1917790,
    "type": "heal",
  },
]
`;
